Schema Configuration documentation (#3349)
neo4j/graphql: docs/modules/ROOT/pages/type-definitions/schema-configuration/index.adoc

[[type-definitions-schema-configuration]]
= Schema Configuration

Neo4j GraphQL Library supports several functionalities such as CRUD operations, aggregation, filtering, and others.
To make them work, a large amount of GraphQL types are generated.

For instance, starting from the type definitions:

[source, graphql, indent=0]
----
type Movie { 
    title: String
    length: Int
} 
----

In some cases, it may be advisable to reduce the scope of the API produced. Whatever the reason is to reduce the size of the schema, improve the performance, or limit access to unwanted operations, this section provides all the information needed to configure the API produced.

> From the type definitions above, let's say we want to prevent users from querying `movies` in the Neo4j Database. To do that, let's bring some changes to the initial type definitions:

Reviewer: ```suggestion
Consider the previously mentioned type definitions.
If you want to prevent users from querying `movies` in the Neo4j Database, you need to make the following the changes:
```

[source, graphql, indent=0]
----
type Movie @query(read: false, aggregate: false) { 
    title: String
    length: Int
} 
----

With this change, the library does **not** generate the following operation fields from the `Query` type:

* `movies`
* `moviesAggregate`
* `moviesConnection`

> NOTE: Please note that aggregations will no longer be generated by default in the 4.x version of the library. See xref::guides/v4-migration/index.adoc#opt-in-aggregation[`Opt-in Aggregation`]

Reviewer: In other places you are using 4.0.0 instead of 4.x when speaking about the version. Is there a more correct or preferred way to refer to the version?

```suggestion
[NOTE]
====
Aggregations will no longer be generated by default in the 4.x version of the library. See xref::guides/v4-migration/index.adoc#opt-in-aggregation[`Opt-in Aggregation`] for more information.
====
```

Author: I would say 4.0.0!

> [[type-definitions-schema-configuration]]

Reviewer: I have the feeling that this page should have the content of type-definitions/schema-configuration.adoc instead. Also, this page is not listed in the content-nav.adoc, which makes me think there was some confusion here as the content is more or less the same.

Author: The other way around, while playing with the folder structure I forget to delete the previous attempt which is the file type-definitions/schema-configuration.adoc, my bad!

= Schema Configuration

Neo4j GraphQL Library supports several functionalities like CRUD operations, aggregation, filtering, and many others.

To make them work, a large amount of GraphQL types are generated.

For instance, starting from the type definitions:

[source, graphql, indent=0]
----
type Movie { 
    title: String
    length: Int
} 
----

> In some cases, it may be advisable to reduce the scope of the API produced. Whatever the reason is to reduce the size of the schema, improve the performance, or limit access to unwanted operations, this section provides all the information needed to configure the API produced.

Reviewer: This is somewhat duplicate content from type-definitions/schema-configuration.adoc but there are a few differences. For instance, there it says "in most cases" instead of "some cases". Also, it's advisable not to have the same content in different pages as this can make Google indexing worse.

> From the type definitions above, let's say we want to prevent users from querying `movies` in the Neo4j Database. To do that, let's bring some changes to the initial type definitions:

Reviewer: There are no type definitions above, but I suppose this was copied from the other file, right? Please let me know if that's the case so I can review this page again :)

[[type-definitions-schema-configuration]]
= Schema Configuration

Neo4j GraphQL Library supports several functionalities such as CRUD operations, aggregation, filtering, and others.
To make them work, a large amount of GraphQL types are generated.

In some cases, it may be advisable to reduce the scope of the API produced. 

> This section provides information on how to reduce the size of the schema, improve the performance, and limit access to unwanted operations.

Reviewer 2: I'm slightly nervous about this line because I don't want users to expect massive performance gains with this as they aren't likely to be very noticeable unless they extensively use these features on a large schema. What do you think about this new wording?

```suggestion
This section provides information on how to limit access to unwanted operations and reduce the size of the schema which can improve the performance.
```

[[type-definitions-schema-configuration]]
= Schema Configuration

Neo4j GraphQL Library supports several functionalities such as CRUD operations, aggregation, filtering, and others.
To make them work, a large amount of GraphQL types are generated.

For instance, starting from the type definitions:

[source, graphql, indent=0]
----
type Movie { 
    title: String
    length: Int
} 
----

In some cases, it may be advisable to reduce the scope of the API produced. 
This section provides information on how to reduce the size of the schema, improve the performance, and limit access to unwanted operations.

Consider the previously mentioned type definitions.

> If you want to prevent users from querying `movies` in the Neo4j Database, you need to make the following the changes:

Author: I personally agree, @lidiazuin , do you have opinions on that?

> If you want to prevent users from querying `movies` in the Neo4j Database, you need to make the following the changes:

Reviewer: Makes sense!
